### ISSUE TITLE
Tests failing on master in CI for a week

### ISSUE THREAD

**Issue by jayvdb** [storyscript/cli#288]
https://circleci.com/gh/storyscript/cli/1004 shows `F` and `E` in the output, and the output of the test runner is all screwed up.  Suspect some threading problem.

When I run the tests, the dots keep happening after tox has finished!

All master builds since then have had similar problems.

https://github.com/storyscript/cli/pull/270 looks like the cause of it, as the previous build looks good: https://circleci.com/gh/storyscript/cli/987

**Comment by jayvdb** [comment]
The problem is the apps "integration" tests - disabling those makes everything lovely again, so this is an immediate/critical side effect of the more general problem https://github.com/storyscript/cli/issues/277

The customing patching and weird fixtures in these Storyscript Python test suites is quite astounding.  There are lots of good pytest plugins, which have their own test suites to verify they work correctly.

And obviously nobody was not paying attention to the test runner output when developing & merging, because it is very visibly broken.

**Comment by jayvdb** [comment]
It looks release 0.20.0+ are all after this, so would be good to get early confirmation that this problem is within the test suite only (very likely), and not the product.  Otherwise this becomes quite critical.

**Comment by wilzbach** [comment]
I think it's just a testsuite problem. There have been no reports about a broken CLI.

**Comment by williammartin** [comment]
> When I run the tests, the dots keep happening after tox has finished!

I also see this. Quite annoying. Interestingly, when they _finally_ "finish", my tests actually do fail (but they don't exit non-zero, and they do absolutely jam up my git repo with coverage files)

**Comment by williammartin** [comment]
This behaviour is almost certainly related to the interaction of pytest and this code: https://github.com/storyscript/cli/blob/master/story/cli.py#L75-L108

I don't fully understand the interaction of the `fork` here but I'm very suspicious of the behaviour. What's going on is that we have two pytest processes running, the first is exiting successfully and the second is failing. This can be reproduced by changing the `tox.ini` command to:

```
    pytest -s -k test_init
```

and running `tox`:

```
± |master {1} U:2 ✗| → tox
GLOB sdist-make: /Users/user/workspace/storyscript/cli/setup.py
py37 inst-nodeps: /Users/user/workspace/storyscript/cli/.tox/.tmp/package/1/story-0.21.1.dev12+gfc93268.d20191007.zip
py37 installed: appdirs==1.4.3,atomicwrites==1.3.0,attrs==19.2.0,blindspin==2.0.1,bom-open==0.4,cachetools==3.1.0,certifi==2019.9.11,chardet==3.0.4,Click==7.0,click-aliases==1.0.1,click-completion==0.5.1,click-help-colors==0.5,click-spinner==0.1.8,coverage==4.5.4,emoji=
=0.5.0,idna==2.8,Jinja2==2.10.3,lark-parser==0.7.2,MarkupSafe==1.1.1,more-itertools==7.2.0,peewee==3.9.3,pluggy==0.6.0,py==1.8.0,pytest==3.6.3,pytest-asyncio==0.8.0,pytest-cov==2.5.1,pytest-mock==1.10.0,pytz==2019.3,PyYAML==5.1.2,raven==6.9.0,requests==2.22.0,semver==2.
8.1,shellingham==1.3.1,six==1.12.0,story==0.21.1.dev12+gfc93268.d20191007,story-hub==0.1.6,storyscript==0.24.2,texttable==1.4.0,urllib3==1.25.6,websockets==7.0,xerox==0.4.1
py37 run-test-pre: PYTHONHASHSEED='1927020353'
py37 run-test: commands[0] | pytest -s -k test_init
======================================================== test session starts =========================================================
platform darwin -- Python 3.7.4, pytest-3.6.3, py-1.8.0, pluggy-0.6.0
rootdir: /Users/user/workspace/storyscript/cli, inifile: pytest.ini
plugins: mock-1.10.0, cov-2.5.1, asyncio-0.8.0
collected 332 items / 330 deselected

tests/integration/commands/apps.py ..

============================================== 2 passed, 330 deselected in 0.41 seconds ==============================================
pep8 inst-nodeps: /Users/user/workspace/storyscript/cli/.tox/.tmp/package/1/story-0.21.1.dev12+gfc93268.d20191007.zip
F

============================================================== FAILURES ==============================================================
_____________________________________________________________ test_init ______________________________________________________________

runner = <MagicMock id='4472599184'>, patch = <pytest_mock.MockFixture._Patcher object at 0x10a967150>

    def test_init(runner, patch):
        app_name = 'myapp'
        patch.object(api.Apps, 'get_uuid_from_hostname')
        from story.commands import apps
        patch.object(apps, 'create_story_yaml', side_effect=apps.create_story_yaml)
        with runner.runner.isolated_filesystem():
            result = runner.run(apps.init, args=[app_name], exit_code=0)
>           with open('story.yml') as f:
E           FileNotFoundError: [Errno 2] No such file or directory: 'story.yml'

tests/integration/commands/apps.py:258: FileNotFoundError
========================================= 1 failed, 1 passed, 330 deselected in 0.99 seconds =========================================
```

`ps` will show that there are two `pytest` processes running, if caught at the right time.

Separately, this can be resolved by adding `TOXENV` to the `passenv` in the `tox.ini` and running:

```
± |master {1} U:2 ✗| → TOXENV=py37 tox
GLOB sdist-make: /Users/user/workspace/storyscript/cli/setup.py
py37 inst-nodeps: /Users/user/workspace/storyscript/cli/.tox/.tmp/package/1/story-0.21.1.dev12+gfc93268.d20191007.zip
py37 installed: appdirs==1.4.3,atomicwrites==1.3.0,attrs==19.2.0,blindspin==2.0.1,bom-open==0.4,cachetools==3.1.0,certifi==2019.9.11,c
hardet==3.0.4,Click==7.0,click-aliases==1.0.1,click-completion==0.5.1,click-help-colors==0.5,click-spinner==0.1.8,coverage==4.5.4,emoj
i==0.5.0,idna==2.8,Jinja2==2.10.3,lark-parser==0.7.2,MarkupSafe==1.1.1,more-itertools==7.2.0,peewee==3.9.3,pluggy==0.6.0,py==1.8.0,pyt
est==3.6.3,pytest-asyncio==0.8.0,pytest-cov==2.5.1,pytest-mock==1.10.0,pytz==2019.3,PyYAML==5.1.2,raven==6.9.0,requests==2.22.0,semver
==2.8.1,shellingham==1.3.1,six==1.12.0,story==0.21.1.dev12+gfc93268.d20191007,story-hub==0.1.6,storyscript==0.24.2,texttable==1.4.0,ur
llib3==1.25.6,websockets==7.0,xerox==0.4.1
py37 run-test-pre: PYTHONHASHSEED='1298613440'
py37 run-test: commands[0] | pytest -s -k test_init
======================================================== test session starts =========================================================
platform darwin -- Python 3.7.4, pytest-3.6.3, py-1.8.0, pluggy-0.6.0
rootdir: /Users/user/workspace/storyscript/cli, inifile: pytest.ini
plugins: mock-1.10.0, cov-2.5.1, asyncio-0.8.0
collected 332 items / 330 deselected

tests/integration/commands/apps.py ..

============================================== 2 passed, 330 deselected in 0.40 seconds ==============================================
______________________________________________________________ summary _______________________________________________________________
  py37: commands succeeded
  congratulations :)
```

We could take time to dig into this weird `fork` behaviour further but I'm pretty sure we're going to need to remove it altogether due to https://github.com/storyscript/cli/issues/280#issuecomment-539057943

**Comment by williammartin** [comment]
Oh, now I'm thinking that the CLI Runner test utility probably captures the `SystemExit` from the tracking code: https://github.com/storyscript/cli/blob/master/story/cli.py#L108

From the `sys.exit()` docs:

```
Exit the interpreter by raising SystemExit(status).
```

And then in the click testing library, this is caught: https://github.com/pallets/click/blob/master/click/testing.py#L325-L327

After this, the program execution will continue, falling back to the `pytest` workload.

**Comment by williammartin** [comment]
The reason that other tests don't fail is that they patch the `track` function https://github.com/storyscript/cli/blob/master/tests/integration/commands/apps.py#L135

**Comment by williammartin** [comment]
Fixed in #308
